src/actix/api/snapshot_api.rs

```diff
@@ -647,7 +648,100 @@ async fn recover_partial_snapshot(
 
         Ok(())
     })
-    .map(|x| x.map_err(Into::into).and_then(|x| x));
+    .map(|res| res.map_err(Into::into).and_then(|res| res));
+
+    helpers::time_or_accept(future, wait.unwrap_or(true)).await
+}
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct PartialSnapshotRecoverFrom {
+    peer_url: Url,
+    api_key: Option<String>,
+}
+
+#[post("/collections/{collection}/shards/{shard}/snapshot/partial/recover_from")]
+async fn recover_partial_snapshot_from(
+    dispatcher: web::Data<Dispatcher>,
+    http_client: web::Data<HttpClient>,
+    path: web::Path<(String, ShardId)>,
+    query: web::Query<SnapshottingParam>,
+    web::Json(request): web::Json<PartialSnapshotRecoverFrom>,
+    ActixAccess(access): ActixAccess,
+) -> impl Responder {
+    let (collection_name, shard_id) = path.into_inner();
+    let PartialSnapshotRecoverFrom { peer_url, api_key } = request;
+    let SnapshottingParam { wait } = query.into_inner();
+
+    // nothing to verify
+    let pass = new_unchecked_verification_pass();
+
+    let future = cancel::future::spawn_cancel_on_drop(async move |cancel| {
+        let cancel_safe = async {
+            let toc = dispatcher.toc(&access, &pass);
+
+            let collection_pass = access
+                .check_global_access(AccessRequirements::new().manage())?
+                .issue_pass(&collection_name)
+                .into_static();
+
+            let collection = toc.get_collection(&collection_pass).await?;
+            collection.assert_shard_exists(shard_id).await?;
+
+            let http_client = http_client.client(api_key.as_deref())?;
+
+            let create_snapshot_url = format!(
+                "{peer_url}/collections/{collection_name}/shards/{shard_id}/snapshot/partial/create"
+            );
+
+            let snapshot_manifest = collection.get_partial_snapshot_manifest(shard_id).await?;
+
+            let download_dir = toc.optional_temp_or_snapshot_temp_path()?;
+            let (partial_snapshot_file, partial_snapshot_temp_path) = tempfile::Builder::new()
+                .prefix("partial-snapshot")
+                .suffix(".download")
+                .tempfile_in(&download_dir)?
+                .into_parts();
+
+            let response = http_client
+                .post(create_snapshot_url)
+                .json(&snapshot_manifest)
+                .send()
+                .await?
+                .error_for_status()?;
+
+            let mut partial_snapshot_file = tokio::fs::File::from_std(partial_snapshot_file);
+            let mut partial_snapshot_stream = response.bytes_stream();
+
+            while let Some(chunk) = partial_snapshot_stream.next().await {
+                partial_snapshot_file.write_all(&chunk?).await?;
+            }
```

Comment:
Any idea if we need [buffering](https://docs.rs/tokio/latest/tokio/io/struct.BufWriter.html) here? I imagine that chunks can be very small.

Reply:
```suggestion
            let mut partial_snapshot_file = tokio::io::BufWriter::new(tokio::fs::File::from_std(partial_snapshot_file));
            let mut partial_snapshot_stream = response.bytes_stream();

            while let Some(chunk) = partial_snapshot_stream.next().await {
                partial_snapshot_file.write_all(&chunk?).await?;
            }
```